### ISSUE TITLE
Add support for .delay_on_commit()

### ISSUE THREAD

**Issue by seth-kerner** [maciej-gol/tenant-schemas-celery#133]
As of Celery version 5.4, `.delay_on_commit()` was introduced as a shortcut function to allow starting a task only after related database transactions are committed (as opposed to the usual `.delay()` function).  See documentation here: https://docs.celeryq.dev/en/stable/django/first-steps-with-django.html#trigger-tasks-at-the-end-of-the-database-transaction

I'm going through the process of updating my django app to use `tenant-schemas-celery` instead of vanilla celery, and I noticed that my tasks which call `.delay_on_commit()` fail with an error that that attribute was not found.  If I traced everything correctly, it appears the issue is because `TenantTask` inherits directly from celery's `Task`, instead of inheriting from `DjangoTask` (located at `celery.contrib.django.task`) as recommended by celery's documentation for any custom task base classes.  Therefore, `.delay_on_command()` no longer exists on task objects.

According to the source documentation page (https://docs.celeryq.dev/en/stable/_modules/celery/contrib/django/task.html#DjangoTask), `DjangoTask` inherits from `Task`, so I'd imagine there should be no major unintended side-effects.  

Potential workarounds:
- Manually perform the functionality by importing `django.db.transaction` and doing `transaction.on_commit(lambda: my_task.delay(args))`
- Create my own base task class (eg. `DjangoTenantTask`) and inherit both `TenantTask` and `DjangoTask`, then use `DjangoTenantTask` in my task's `shared_task` decorator function

May I request that `TenantTask` inherits from `DjangoTask` to avoid the workarounds?

**Comment by maciej-gol** [comment]
Thanks, will look into it!

pon., 23 gru 2024, 18:23 użytkownik seth-kerner ***@***.***>
napisał:

**Comment by maciej-gol** [comment]
I have just added a new task class `TenantDjangoTask` that subclasses `DjangoTask`. This is required because the package has to support celery<5.4, too.

**Comment by seth-kerner** [comment]
Thanks!  I'm looking forward to all upcoming enhancements in the 4.0.0 release!
